src/ipset/ipset_hash.c

```diff
@@ -0,0 +1,345 @@
+/*
+ * DPVS is a software load balancer (Virtual Server) based on DPDK.
+ *
+ * Copyright (C) 2021 iQIYI (www.iqiyi.com).
+ * All Rights Reserved.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ */
+#include "conf/common.h"
+#include "ipset/ipset_hash.h"
+#include "ipset/pfxlen.h"
+
+#define DEF_HASHSIZE    1024
+#define DEF_MAXELEM     65535
+
+#define data_equal(e1, e2) set->variant->hash.data_equal(e1, e2)
+#define data_netmask(v, c, b) set->variant->hash.data_netmask(v, c, b)
+#define hashkey(d, l, m) set->variant->hash.hashkey(d, l, m)
+
+/* hash mempool */
+#define IPSET_HASH_POOL_SIZE_DEF    65536
+#define IPSET_HASH_CACHE_SIZE_DEF   256
+/* this should be larger than the element of all types */
+#define HASH_ELEM_SIZE              128
+#define this_hash_cache             (ipset_hash_cache[rte_socket_id()])
+
+static struct rte_mempool *ipset_hash_cache[DPVS_MAX_SOCKET];
+
+/* common hash element difinition */
+typedef struct hash_elem {
+    union inet_addr ip1;
+    uint8_t cidr;
+    union inet_addr ip2;
+    uint8_t cidr2;
+} elem_t;
+
+void
+hash_data_netmask4(void *elem, uint8_t cidr, bool inner)
+{
+    elem_t *e = (elem_t *)elem;
+
+    if (inner) {
+        e->ip2.in.s_addr &= ip_set_netmask(cidr);
+        e->cidr2 = cidr;
+    } else {
+        e->ip1.in.s_addr &= ip_set_netmask(cidr);
+        e->cidr = cidr;
+    }
+}
+
+void
+hash_data_netmask6(void *elem, uint8_t cidr, bool inner)
+{
+    elem_t *e = (elem_t *)elem;
+
+    if (inner) {
+        ip6_netmask(&e->ip2, cidr);
+        e->cidr2 = cidr;
+    } else {
+        ip6_netmask(&e->ip1, cidr);
+        e->cidr = cidr;
+    }
+}
+
+uint32_t
+jhash_hashkey(void *data, int len, uint32_t mask)
+{
+    return rte_jhash(data, len, 0) & mask;
+}
+
+static int
+hash_add(struct ipset *set, void *value)
+{
+    struct hash_type *htype = set->data;
+    struct hash_entry *hnode;
+    struct list_head *head;
+    uint32_t key;
+    void *obj, *elem;
+    elem_t *e;
+
+    if (unlikely(set->elements >= htype->maxelem))
+        return EDPVS_NOMEM;
+
+    key = hashkey(value, set->hash_len, htype->mask);
+    head = &htype->htable[key];
+    list_for_each_entry(hnode, head, list) {
+        if (data_equal(value, hnode->elem)) {
+            return EDPVS_EXIST;
+        }
+    }
+
+    /* obj memory layout
+       | hnode | elem | */
+    rte_mempool_get(this_hash_cache, &obj); 
+    memset(obj, 0,  sizeof(struct hash_entry) + HASH_ELEM_SIZE);
```

Comment:
Check the return value of `rte_mempool_get` to ensure the memory pointed by `obj`  is safe.